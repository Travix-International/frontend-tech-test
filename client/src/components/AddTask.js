import React, { PureComponent } from 'react';

import { connect } from 'react-redux';
import { createTask } from '../actions/tasks';

import Modal from './Modal';
import TaskForm from './TaskForm';
import { btnTextAdd, modalTitleAdd } from '../constants';

import '../style/AddTask.scss';

class AddTask extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      showModal: false
    };
    this.openModal = this.openModal.bind(this);
    this.closeModal = this.closeModal.bind(this);
    this.addTask = this.addTask.bind(this);
  }

  openModal(task) {
    return this.setState({ showModal: true });
  }

  closeModal() {
    this.setState({ showModal: false });
  }

  addTask(taskDetails) {
    const { createTask } = this.props;
    createTask(taskDetails);
    this.closeModal();
  }

  render() {
    return (
      <div className="AddTask">
        <button
          className="cta-btn"
          onClick={this.openModal}>add task
        </button>
        <Modal
          modalTitle={modalTitleAdd}
          showModal={this.state.showModal}
          closeModal={this.closeModal}>
          <TaskForm
            btnText={btnTextAdd}
            onSubmit={this.addTask} />
        </Modal>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    createTask: (taskDetails) => {
      dispatch(createTask(taskDetails));
    }
  };
};

const ConnectedAddTask = connect(
  undefined,
  mapDispatchToProps
)(AddTask);

export default ConnectedAddTask;
