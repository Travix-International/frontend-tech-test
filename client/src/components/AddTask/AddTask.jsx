import React, {useState, useEffect} from 'react';
import PropTypes from 'prop-types';

import styles from './AddTask.module.scss';
import Input from '../UI/Input/Input';
import Button from '../UI/Button/Button';
import TextArea from '../UI/TextArea/TextArea';
import { TaskType } from '../../constants/propTypes';

const AddTask = ({editMode, handleAddTask, editData, handleClose, handleEditTask}) => {
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');

  useEffect(() => {
    if(editData){
      setTitle(decodeURI(editData.title));
      setDescription(decodeURI(editData.description));
    }
  }, [editMode])

  const handleSubmit = (event) => {
    event.preventDefault();
    if(!title || !description) {
      alert('please fill all inputs');
      return;
    }
    if(editMode !== false) {
      handleEditTask(editData.id, title, description, handleClose);
    } else {
      handleAddTask(title, description, handleClose);
    }
  }

  return(
    <form className={styles.AddTask} onSubmit={handleSubmit}>
      <div className={styles.Row}>
        <Input type="text" value={title} onChange={(event) => setTitle(event.target.value)} placeholder="Title" />
      </div>
      <div className={styles.Row}>
        <TextArea value={description} onChange={(event) => setDescription(event.target.value)} placeholder="Description" />
      </div>
      <div className={`${styles.Row} ${styles.Button}`}>
        <Button type="submit">
          {editMode ? "Edit" : "Add"} Task
        </Button>
      </div>
    </form>
  )
}

AddTask.propTypes = {
  handleAddTask: PropTypes.func.isRequired,
  editMode: PropTypes.oneOfType([PropTypes.number, PropTypes.bool]),
  editData: TaskType,
  handleClose: PropTypes.func.isRequired,
  handleEditTask: PropTypes.func.isRequired
}

AddTask.defaultProps = {
  editMode: false,
  editData: {
    id: 0,
    title: '',
    description: ''
  }
}

export default AddTask;