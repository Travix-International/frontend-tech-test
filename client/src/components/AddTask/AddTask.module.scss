.AddTask {
  display: flex;
  flex-direction: column;
  width: 100%;

  .Row {
    margin-top: 10px;
  }

  .Input {
    margin-right: 30px;
    width: 420px;
  }

  .Button {
    display: flex;
    justify-content: center;
  }
}