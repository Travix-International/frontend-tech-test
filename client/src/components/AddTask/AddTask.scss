.add-task {
  background-color: #ffffff;

  .form-group {
    padding: 0 1rem 1rem 1rem;
  }

  h3 {
    cursor: pointer;
    color: #3d91ff;
  }

  hr {
    border: 1px solid #ddd;
    margin-bottom: 1rem;
  }

  .task-button {
    display: flex;
    align-items: center;
    padding: 1rem;
    cursor: pointer;
  }

  .form-control {
    display: grid;
    grid-template-columns: 20% 40%;
    margin-bottom: 1rem;

    label {
      margin: auto 0;
    }
  }
  .icon-plus {
    background-image: url("data:image/svg+xml,%3C%3Fxml version='1.0' %3F%3E%3C!DOCTYPE svg PUBLIC '-//W3C//DTD SVG 1.1//EN' 'http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd'%3E%3Csvg height='128px' id='Layer_1' fill='%233d91ff' style='enable-background:new 0 0 512 512;' version='1.1' viewBox='0 0 512 512' width='128px' xml:space='preserve' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'%3E%3Cg%3E%3Cg%3E%3Cg%3E%3Cpath d='M256,48C141.1,48,48,141.1,48,256s93.1,208,208,208c114.9,0,208-93.1,208-208S370.9,48,256,48z M256,446.7 c-105.1,0-190.7-85.5-190.7-190.7S150.9,65.3,256,65.3S446.7,150.9,446.7,256S361.1,446.7,256,446.7z'/%3E%3C/g%3E%3C/g%3E%3Cg%3E%3Cpolygon points='264.1,128 247.3,128 247.3,247.9 128,247.9 128,264.7 247.3,264.7 247.3,384 264.1,384 264.1,264.7 384,264.7 384,247.9 264.1,247.9 '/%3E%3C/g%3E%3C/g%3E%3C/svg%3E");
    height: 1.5rem;
    width: 1.5rem;
    background-size: cover;
  }

  &:hover {
    box-shadow: 0 0 6px 2px #ddd;
  }
}
