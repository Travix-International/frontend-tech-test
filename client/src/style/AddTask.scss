@import "colors";
@import "mixins";

.AddTask {
	.cta-btn {
		@include button;
		}
	}
