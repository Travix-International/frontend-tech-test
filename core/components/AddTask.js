import React, { Component } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types';
import classNames from 'classnames';

class AddTask extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeAddTodo: false,
      title: '',
      description: '',
      formHasError: false,
    };
    this.handleAddTaskFormChange = this.handleAddTaskFormChange.bind(this);
    this.handleFocusAddTodo = this.handleFocusAddTodo.bind(this);
    this.submitAddTaskForm = this.submitAddTaskForm.bind(this);
    this.handleAddTaskFormChange = this.handleAddTaskFormChange.bind(this);
  }

  handleFocusAddTodo() {
    this.setState({
      activeAddTodo: true,
      formHasError: false,
    });
  }

  handleAddTaskFormChange(event) {
    const { target } = event;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const { name } = target;
    this.setState({
      [name]: value,
      activeAddTodo: true,
      formHasError: false,
    });
  }

  submitAddTaskForm(event) {
    event.preventDefault();
    const { title, description } = this.state;
    const { handleAddTask } = this.props;
    if (title && title.length > 0) {
      handleAddTask({ title, description });
      this.setState({
        title: '',
        description: '',
        formHasError: false,
        activeAddTodo: false,
      });
    } else {
      this.setState({ formHasError: true });
    }
  }

  render() {
    const {
      activeAddTodo, title, description, formHasError,
    } = this.state;
    const addTaskFromClasses = classNames({
      active: activeAddTodo,
      'has-error': formHasError,
    });

    return (
      <form className={addTaskFromClasses} id="add-todo" name="addTask" onSubmit={this.submitAddTaskForm}>
        <div className="title-div">
          <FontAwesomeIcon
            className="add-icon"
            icon={faPlus}
            onClick={this.submitAddTaskForm}
          />
          <input
            autoComplete="off"
            className="add-todo-input"
            name="title"
            onChange={this.handleAddTaskFormChange}
            onFocus={this.handleFocusAddTodo}
            placeholder="Add to-do ..."
            type="text"
            value={title}
          />
        </div>

        <textarea
          autoComplete="off"
          className="description"
          name="description"
          onChange={this.handleAddTaskFormChange}
          placeholder="Add a description!!"
          value={description}
        />
      </form>
    );
  }
}
AddTask.propTypes = {
  handleAddTask: PropTypes.func.isRequired,
};
export default AddTask;
