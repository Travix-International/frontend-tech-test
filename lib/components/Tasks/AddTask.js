import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import * as taskActions from '../../actions/taskActions';

class AddTask extends Component {
  static propTypes = {
    actions: PropTypes.object.isRequired
  };

  state = {
    text: ''
  };

  handleSubmit = (event) => {
    event.preventDefault();

    const { text } = this.state;

    if (text.trim().length) {
      this.props.actions.createTask({
        title: text,
        description: 'Yet another task'
      });
      this.setState({
        text: ''
      });
    } else {
      // better UX by focusing on the input with the error
      this.todoText.focus();
    }
  };

  handleChange = (event) => {
    this.setState({
      text: event.target.value
    });
  };

  render() {
    /* eslint-disable no-return-assign */
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          id="todo-text"
          name="todo-text"
          onChange={this.handleChange}
          placeholder="something you doing?"
          ref={node => (this.todoText = node)}
          type="text"
          value={this.state.text}
        />
        <input type="submit" />
      </form>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(taskActions, dispatch)
});

export default connect(null, mapDispatchToProps)(AddTask);
