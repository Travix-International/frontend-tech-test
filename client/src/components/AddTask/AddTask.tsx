import React, { useState } from "react";

import { Task } from "../../models/Task";
import "./AddTask.scss";

interface Props {
  change: (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void;
  add: () => void;
  task: Task;
}

const AddTask: React.FC<Props> = ({ change, add, task }) => {
  const [expand, setExpand] = useState(false);

  return (
    <div className="add-task">
      <div onClick={() => setExpand(!expand)} className="task-button">
        <div className="icon-plus"></div>
        <h3>Add new task</h3>
      </div>
      {expand && (
        <div className="form-group">
          <hr />
          <form className="form-control">
            <label htmlFor="title">Title: </label>
            <input
              type="text"
              name="title"
              id="title"
              onChange={change}
              value={task.title}
              required
            />
          </form>

          <div className="form-control">
            <label htmlFor="category">Category:</label>
            <input
              type="text"
              name="category"
              id="category"
              onChange={change}
              value={task.category}
              required
            />
          </div>

          <div className="form-control">
            <label htmlFor="description">Description:</label>
            <textarea
              name="description"
              id="description"
              onChange={change}
              value={task.description}
              required
            />
          </div>

          <button type="submit" onClick={add}>
            Add
          </button>
        </div>
      )}
    </div>
  );
};

export default AddTask;
