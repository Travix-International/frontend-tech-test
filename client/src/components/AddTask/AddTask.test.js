import React from 'react';
import { shallow } from 'enzyme';
import AddTask from './AddTask';

it('renders AddTask correctly', () => {
  const wrapper = shallow(
  <AddTask 
    handleAddTask={() => {}} 
    handleClose={() => {}}
    handleEditTask={() => {}} />);

  expect(wrapper).toMatchSnapshot();
});