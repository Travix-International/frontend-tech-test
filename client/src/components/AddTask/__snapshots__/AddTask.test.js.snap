// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AddTask correctly 1`] = `ShallowWrapper {}`;
